.NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs

```diff
@@ -1427,40 +1420,22 @@ private DateTimeResolutionResult ParseWeekOfYear(string text, DateObject referen
             }
 
             DateObject targetWeekMonday;
-            int weekNum;
+
             if (this.config.IsLastCardinal(cardinalStr))
             {
-                var lastDay = DateObject.MinValue.SafeCreateFromValue(year, 12, 31);
-                var lastDayWeekThursday = lastDay.This(DayOfWeek.Thursday);
-                weekNum = Cal.GetWeekOfYear(lastDayWeekThursday, CalendarWeekRule.FirstFourDayWeek, DayOfWeek.Monday);
-
-                // Thursday fall into next year's first week
-                if (weekNum == 1)
-                {
-                    lastDayWeekThursday = lastDayWeekThursday.AddDays(-Constants.WeekDayCount);
-                }
-
-                targetWeekMonday = lastDayWeekThursday.This(DayOfWeek.Monday);
+                targetWeekMonday = GetLastThursday(year).This(DayOfWeek.Monday);
 
                 ret.Timex = TimexUtility.GenerateWeekTimex(targetWeekMonday);
             }
             else
             {
-                var firstDay = DateObject.MinValue.SafeCreateFromValue(year, 1, 1);
-                DateObject firstDayWeekThursday = firstDay.This(DayOfWeek.Thursday);
-                weekNum = Cal.GetWeekOfYear(firstDay, CalendarWeekRule.FirstFourDayWeek, DayOfWeek.Monday);
-
-                // Thursday fall into previous year's last week
-                if (weekNum != 1)
-                {
-                    firstDayWeekThursday = firstDay.AddDays(Constants.WeekDayCount);
-                }
-
                 var cardinal = this.config.CardinalMap[cardinalStr];
-                targetWeekMonday = firstDayWeekThursday.This(DayOfWeek.Monday)
+                targetWeekMonday = GetFirstThursday(year).This(DayOfWeek.Monday)
                     .AddDays(Constants.WeekDayCount * (cardinal - 1));
 
-                ret.Timex = TimexUtility.GenerateWeekTimex(targetWeekMonday);
+                var weekTimex = TimexUtility.GenerateWeekTimex(cardinal);
+                var yearTimex = DateTimeFormatUtil.LuisDate(year);
```

Comment:
This should be a single timex generation method that receives the necessary input and returns the final timex.